### ISSUE TITLE
Have support for updating specific pages, to give true rectangular boundary control of updated regions --> Higher framerate:)

### ISSUE THREAD

**Issue by joostit** [samveen/pico-oled-1.3-driver#8]
To enable higher framerates, it's useful to not only update specific columns (X-Axis in datasheet) of the display, but to also do this for pages (Y-Axis in datasheet). Doing so enables updating very specific rectangular bounds on the display. This will greatly increase framerate for that specific region and for the display as a whole.

I already have this working and will prepare a pull request in the coming days (i.e. clean up my code and make it more robust ;) )

**Current Situation:** `def show(self, start=0, end=64)`
**Proposed situation:** `def show(self, startCol=0, endCol=64, startPage=0, endPage=16)`
_Note that the display only supports updating whole pages (8 pixels) only, hence there's 16 pages._

Thoughts are welcome :)

**Comment by joostit** [comment]
A second proposal is not to 'silently try to fix' invalid parameters in the Show(...) method. In my experience this will hide bugs in other parts of the application and make debugging difficult. In my code I raise IndexErrors when wrong parameter values are given. I propose to do the same for the the existing startCol and endCol parameters.

**Comment by samveen** [comment]
> To enable higher framerates, it's useful to not only update specific columns (X-Axis in datasheet) of the display, but to also do this for pages (Y-Axis in datasheet). Doing so enables updating very specific rectangular bounds on the display. This will greatly increase framerate for that specific region and for the display as a whole.

That's been on my mind too (see  https://forums.raspberrypi.com/viewtopic.php?p=2210815#p2210815 ) (and my TODO list). However the height improvements are by @mungewell , so it will probably take me some time to dig into how to fix this. 

> I already have this working and will prepare a pull request in the coming days (i.e. clean up my code and make it more robust ;) )


What are you making me wait for? Hurry up :rofl: !! @mungewell  already has some changes he's implemented, but not yet pushed to this repo at https://github.com/samveen/pico-oled-1.3-driver/commit/073a5e9cc9c440129019a0b259c8e210bc586a9d . Please take a look at it too. 

> **Current Situation:** `def show(self, start=0, end=64)` **Proposed situation:** `def show(self, startCol=0, endCol=64, startPage=0, endPage=16)` _Note that the display only supports updating whole pages (8 pixels) only, hence there's 16 pages._

Can this be modified into `def show(self, (start_x,start_y), (end_x,end_y))` , where x and y are row and column coords, instead of pages? I am hoping that the value of pages could possibly be derived from the start and end coordinates of the rectangle that needs to be updated. This also ends up fitting well with the framebuffer drawing class that these displays are derived from.

> A second proposal is not to 'silently try to fix' invalid parameters in the Show(...) method. In my experience this will hide bugs in other parts of the application and make debugging difficult. In my code I raise IndexErrors when wrong parameter values are given. I propose to do the same for the the existing startCol and endCol parameters.

Would you please raise the exception pull request first? :pray:  (See that pun there!!) . I've checked in some minor improvements that I was testing. Please rebase your work on that if possible.

**Comment by joostit** [comment]
> 
> What are you making me wait for? Hurry up 🤣 !! @mungewell already has some changes he's implemented, but not yet pushed to this repo at [073a5e9](https://github.com/samveen/pico-oled-1.3-driver/commit/073a5e9cc9c440129019a0b259c8e210bc586a9d) . Please take a look at it too.
> 
Haha On it  💪🏻  Looking at the code in the other commit I think it might become obsolete now that we've already got it to work. (or I fail to understand its purpose 😅 )

> Can this be modified into `def show(self, (start_x,start_y), (end_x,end_y))` , where x and y are row and column coords, instead of pages? I am hoping that the value of pages could possibly be derived from the start and end coordinates of the rectangle that needs to be updated. This also ends up fitting well with the framebuffer drawing class that these displays are derived from.

Good idea. In that case I will use the X/Y reference from the frame buffer. **Note that this will be a breaking change with existing code as the parameters for show() change** (which I find the reasonable thing to do here). Initially I will only give page-wise control over the X-axis. Later we can see if/how we can do pixel-wise control. At this moment I doubt whether it's possible to efficiently write partial pages to the display. ( `//ToDo: Figure this out` )

> Would you please raise the exception pull request first? 🙏 (See that pun there!!) . I've checked in some minor improvements that I was testing. Please rebase your work on that if possible.

I just made a separate issue for this (#10) and already created a pull request. Once it has been merged, I will create a new pull request for the horizontal update boundaries, to prevent merge conflicts.

**Comment by samveen** [comment]
From what I understand of the code by @mungewell, it does partial page writes, mapping the `start(x,y), end(x,y)` (`start`,`end` -> `y` , `start_col`, `end_col` -> `x`) into partial page writes. That is why I suggested that you take a look at it.

More comments on the the PR and #10

**Comment by joostit** [comment]
As far as I can see from commit 073a5e9cc9c440129019a0b259c8e210bc586a9d, it 's working towards the same solution as what I did, but keeping the user in the dark about the limitations of the display interface. On line 159 it bitshifts the X-coordinate three bits to the right, ignoring the three lower bits and effectively giving page-wise access anyways.
`for num in range(start_col>>3, (end_col+7)>>3):`

My doubts about that approach is that it obfuscates to the user that they don't really have pixel-based resolution, even though the show() API suggests it. It'll make the display update act differently that the user would expect, probably causing frustration and notebooks being thrown out of windows 😝 
I personally think the lesser of two evils is to be a bit less consistent in the `show()` API (_by using the 8-pixel page index, instead of pure pixels for X_), than to have misleading functionality and cause confusion.

My suggestion would be: _(until we figure out how to have real pixel-based access to the display, if possible at all)_

- Be clear and honest to the user about the X-axis update resolution being page-based, instead of pixel-based. This might confuse users at first, so mitigate that challenge with:
  - Have proper naming for parameters: `show((xStartPage, yStartLine), (xEndPage, yEndLine))`.
  - Have proper comments on top of the `show(...)` method to describe the parameters and their meaning/purpose.
  - Raise proper Exceptions when the user inputs the wrong values for the page numbers, including a helpful message.
  - Provide helpful demo code with explanations. (~~Have a reference to the datasheet and RTFM~~)
    Also being talked about in #5
  - Maybe provide a public helper method `(xStartPage, xEndPage) = XtoPage(xStartPx, xEndPx)`, which makes the given X-pixels fit within updatable pages?


I do find this a difficult decision, consistency vs accuracy. My assumption here is that users who use the more elaborate overload for `show(...)`, will do so for performance reasons. They're probably clever developers anyways, who appreciate accurate control over what they're doing 🤔 

What are your thoughts @samveen

**Comment by samveen** [comment]
> As far as I can see from commit [073a5e9](https://github.com/samveen/pico-oled-1.3-driver/commit/073a5e9cc9c440129019a0b259c8e210bc586a9d), it 's working towards the same solution as what I did, but keeping the user in the dark about the limitations of the display interface. On line 159 it bitshifts the X-coordinate three bits to the right, ignoring the three lower bits and effectively giving page-wise access anyways. `for num in range(start_col>>3, (end_col+7)>>3):`

This is converting the column count from pixels into pages -> `range(0,128)` => `range(0,16)`

> My doubts about that approach is that it obfuscates to the user that they don't really have pixel-based resolution, even though the show() API suggests it. It'll make the display update act differently that the user would expect, probably causing frustration and notebooks being thrown out of windows 😝 I personally think the lesser of two evils is to be a bit less consistent in the `show()` API (_by using the 8-pixel page index, instead of pure pixels for X_), than to have misleading functionality and cause confusion.

The frame-buffer is the screen as far as the user is concerned. However, each screen driver uses a different underlying protocol for how data is to be written to the screen, which is weird and downright convoluted in many cases. These details should ideally be hidden from the user as they'll end up needing to do more background reading, than actually solving the problem they need to solve (you and I are doing exactly that, right). So ideally the show should be consistent with how the other framebuffer functions behave, while magically fixing things in-code.


> My suggestion would be: _(until we figure out how to have real pixel-based access to the d
isplay, if possible at all)_

That's going to be dependent of the actual display chip's I2C data transfer instruction. I've seen the display chip's I2C commmunication protocol details in a PDF somewhere on the WaveShare website, and I was very unhappy with how convoluted it's update implementation is. That is the reason why I ended up using their driver as the base, and improving it to my requirements, instead of  writing something better from scratch.

>       * Raise proper Exceptions when the user inputs the wrong values for the page numbers, including a helpful message.
>       * Provide helpful demo code with explanations. (~Have a reference to the datasheet and RTFM~)

I completely agree on these two. The rest need a larger discussion than just you and me :sweat_smile: 

> I do find this a difficult decision, consistency vs accuracy. My assumption here is that users who use the more elaborate overload for `show(...)`, will do so for performance reasons. They're probably clever developers anyways, who appreciate accurate control over what they're doing 🤔

If the code behaves in a manner consistent with other functions, it'll be easy to use by everyone. Instead of forcing smart people to look under the hood, let's it be something that "just works" as expected.

**Comment by joostit** [comment]
I absolutely agree with you about the user-friendly part. I did have a lot of ~~fun~~ **not fun** trying to understand the display update logic from that datasheet 😅 
For my own goals (training developers), I've decided to go for the more technical solution. The trainees have to puzzle a bit and learn that software under the hood is not always as straightforward as they're sometimes used to. (and learn about the bits and the bytes along the way)

I'll implement this in my Fork and of course give you credits for your awesome work on this driver. Thank you so much! 😄  
I'll consider this issue closed, but feel free to take changes from my Fork in the future if useful :)
